aws/resource_aws_wafregional_rate_based_rule.go

```diff
@@ -0,0 +1,196 @@
+)
+
+func resourceAwsWafRegionalRateBasedRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWafRegionalRateBasedRuleCreate,
+		Read:   resourceAwsWafRegionalRateBasedRuleRead,
+		Update: resourceAwsWafRegionalRateBasedRuleUpdate,
+		Delete: resourceAwsWafRegionalRateBasedRuleDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"metric_name": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateWafMetricName,
+			},
+			"predicate": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"negated": &schema.Schema{
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"data_id": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validateMaxLength(128),
+						},
+						"type": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validateWafPredicatesType(),
+						},
+					},
+				},
+			},
+			"rate_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"rate_limit": &schema.Schema{
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntAtLeast(2000),
+			},
+		},
+	}
+}
+
+func resourceAwsWafRegionalRateBasedRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafregionalconn
+	region := meta.(*AWSClient).region
+
+	wr := newWafRegionalRetryer(conn, region)
+	out, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+		params := &waf.CreateRateBasedRuleInput{
+			ChangeToken: token,
+			MetricName:  aws.String(d.Get("metric_name").(string)),
+			Name:        aws.String(d.Get("name").(string)),
+			RateKey:     aws.String(d.Get("rate_key").(string)),
+			RateLimit:   aws.Int64(int64(d.Get("rate_limit").(int))),
+		}
+
+		return conn.CreateRateBasedRule(params)
+	})
+	if err != nil {
+		return err
+	}
+	resp := out.(*waf.CreateRateBasedRuleOutput)
+	d.SetId(*resp.Rule.RuleId)
+	return resourceAwsWafRegionalRateBasedRuleUpdate(d, meta)
+}
+
+func resourceAwsWafRegionalRateBasedRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafregionalconn
+
+	params := &waf.GetRateBasedRuleInput{
+		RuleId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetRateBasedRule(params)
+	if err != nil {
+		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+			log.Printf("[WARN] WAF Regional Rate Based Rule (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	var predicates []map[string]interface{}
+
+	for _, predicateSet := range resp.Rule.MatchPredicates {
+		predicate := map[string]interface{}{
```

Comment:
I don't know if I can argue it's better but you could avoid the `predicate` variable
```
predicates = append(predicates, map[string]interface{}{
  "negated": *predicateSet.Negated,
  "type":    *predicateSet.Type,
  "data_id": *predicateSet.DataId,
})
```

Reply:
That's a valid argument 👍 I will address that.